module/zfs/arc.c

```diff
@@ -1732,6 +1812,255 @@ arc_buf_try_copy_decompressed_data(arc_buf_t *buf)
+	 * abd so we can verify the MAC. The performance overhead of this will
+	 * be relatively low, since most objects in an encrypted objset will
+	 * be encrypted (instead of authenticated) anyway.
+	 */
+	if (HDR_GET_COMPRESS(hdr) != ZIO_COMPRESS_OFF &&
+	    !HDR_COMPRESSION_ENABLED(hdr)) {
+		tmpbuf = zio_buf_alloc(lsize);
+		abd = abd_get_from_buf(tmpbuf, lsize);
+		abd_take_ownership_of_buf(abd, B_TRUE);
+
+		csize = zio_compress_data(HDR_GET_COMPRESS(hdr),
+		    hdr->b_l1hdr.b_pabd, tmpbuf, lsize);
+		ASSERT3U(csize, <=, psize);
+		abd_zero_off(abd, csize, psize - csize);
+	}
+
+	/*
+	 * Authentication is best effort. We authenticate whenever the key is
+	 * available. If we succeed we clear ARC_FLAG_NOAUTH.
+	 */
+	if (hdr->b_crypt_hdr.b_ot == DMU_OT_OBJSET) {
+		ASSERT3U(HDR_GET_COMPRESS(hdr), ==, ZIO_COMPRESS_OFF);
+		ASSERT3U(lsize, ==, psize);
+		ret = spa_do_crypt_objset_mac_abd(B_FALSE, spa, dsobj, abd,
+		    psize, hdr->b_l1hdr.b_byteswap != DMU_BSWAP_NUMFUNCS);
+	} else {
+		ret = spa_do_crypt_mac_abd(B_FALSE, spa, dsobj, abd, psize,
+		    hdr->b_crypt_hdr.b_mac);
+	}
+
+	if (ret == 0)
+		arc_hdr_clear_flags(hdr, ARC_FLAG_NOAUTH);
+	else if (ret != ENOENT)
+		goto error;
+
+out_unlock:
+	if (tmpbuf != NULL)
+		abd_free(abd);
+	if (hash_lock != NULL)
+		mutex_exit(hash_lock);
+
+	return (0);
+
+error:
+	if (tmpbuf != NULL)
+		abd_free(abd);
+	if (hash_lock != NULL)
+		mutex_exit(hash_lock);
+
+	return (ret);
+}
+
+/*
+ * This function will take a header that only has raw encrypted data in
+ * b_crypt_hdr.b_rabd and decrypts it into a new buffer which is stored in
+ * b_l1hdr.b_pabd. If designated in the header, this function will also
+ * decompress the data.
+ */
+static int
+arc_hdr_decrypt(arc_buf_hdr_t *hdr, kmutex_t *hash_lock, spa_t *spa,
+    uint64_t dsobj)
+{
+	int ret;
+	dsl_crypto_key_t *dck = NULL;
+	abd_t *cabd = NULL;
+	void *tmp = NULL;
+	boolean_t no_crypt = B_FALSE;
+	boolean_t bswap = (hdr->b_l1hdr.b_byteswap != DMU_BSWAP_NUMFUNCS);
+
+	if (hash_lock != NULL)
+		mutex_enter(hash_lock);
+
+	ASSERT(HDR_ENCRYPTED(hdr));
+
+	/*
+	 * Check that we only have an encrypted copy of the data. This was
+	 * already checked in arc_buf_fill() as a quick check, but we do it
+	 * again now under the hash_lock to make sure nothing has changed.
+	 * If this isn't true there is no work to do so we can simply return.
+	 */
+	if (!HDR_HAS_RABD(hdr) || hdr->b_l1hdr.b_pabd != NULL)
+		goto out_unlock;
+
+	arc_hdr_alloc_abd(hdr, B_FALSE);
+
+	/*
+	 * We must be careful to use the passed-in dsobj value here and
+	 * not the value in b_dsobj. b_dsobj is meant to be a best guess for
+	 * the L2ARC, which has the luxury of being able to fail without real
+	 * consequences (the data simply won't make it to the L2ARC). In
+	 * reality, the dsobj stored in the header may belong to a dataset
+	 * that has been unmounted or otherwise disowned, meaning the key
+	 * won't be accessible via that dsobj anymore.
```

Comment:
I went back and forth on this a couple of times. The answer I arrived at is "it depends if you keep your datasets mounted all the time". If you have a dataset that is mounted and you do a `zfs send` of one of its snapshots you'll probably want to keep the key for the mounted dataset instead of the snapshot since the snapshot will not be owned by the end of the send. It seemed more likely to me that the first `arc_read()` would be associated with the mountpoint instead of the send. Really this could probably go either way.